data/beaconrestapi/src/test/java/tech/pegasys/teku/beaconrestapi/handlers/validator/PostAggregateAndProofTest.java

```diff
@@ -0,0 +1,93 @@
+import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
+import static javax.servlet.http.HttpServletResponse.SC_OK;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import io.javalin.http.Context;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+import tech.pegasys.teku.api.ValidatorDataProvider;
+import tech.pegasys.teku.datastructures.operations.SignedAggregateAndProof;
+import tech.pegasys.teku.datastructures.util.DataStructureUtil;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.provider.JsonProvider;
+
+class PostAggregateAndProofTest {
+
+  @SuppressWarnings({"unchecked", "unused"})
+  private final ArgumentCaptor<SafeFuture<String>> args = ArgumentCaptor.forClass(SafeFuture.class);
+
+  @SuppressWarnings("unused")
+  private final DataStructureUtil dataStructureUtil = new DataStructureUtil();
+
+  private final Context context = mock(Context.class);
+  private final ValidatorDataProvider provider = mock(ValidatorDataProvider.class);
+  private final JsonProvider jsonProvider = new JsonProvider();
+
+  private PostAggregateAndProof handler;
+
+  @BeforeEach
+  public void beforeEach() {
+    handler = new PostAggregateAndProof(provider, jsonProvider);
+  }
+
+  @Test
+  public void shouldReturnBadRequestWhenRequestBodyIsInvalid() throws Exception {
+    when(context.body()).thenReturn("{\"foo\": \"bar\"}");
+
+    handler.handle(context);
+    verify(context).status(SC_BAD_REQUEST);
+  }
+
+  @Test
+  public void shouldReturnServerErrorWhenFutureHasUnmappedException() throws Exception {
+    final SignedAggregateAndProof signedAggregateAndProof =
+        dataStructureUtil.randomSignedAggregateAndProof();
+
+    final tech.pegasys.teku.api.schema.SignedAggregateAndProof schemaSignedAggregateAndProof =
+        new tech.pegasys.teku.api.schema.SignedAggregateAndProof(signedAggregateAndProof);
+
+    when(context.body()).thenReturn(jsonProvider.objectToJSON(schemaSignedAggregateAndProof));
+    doThrow(new RuntimeException()).when(provider).sendAggregateAndProof(any());
+
+    handler.handle(context);
+
+    verify(context).status(SC_INTERNAL_SERVER_ERROR);
+  }
+
+  @Test
+  public void shouldReturnSuccessWhenSendAggregateAndProofSucceeds() throws Exception {
+    final SignedAggregateAndProof signedAggregateAndProof =
+        dataStructureUtil.randomSignedAggregateAndProof();
+
+    final tech.pegasys.teku.api.schema.SignedAggregateAndProof schemaSignedAggregateAndProof =
+        new tech.pegasys.teku.api.schema.SignedAggregateAndProof(signedAggregateAndProof);
+
+    when(context.body()).thenReturn(jsonProvider.objectToJSON(schemaSignedAggregateAndProof));
+
+    handler.handle(context);
+
+    verify(provider).sendAggregateAndProof(any());
```

Comment:
we should really either use spy's or use a verify that shows the expected object is coming in... this shows an aggregate got sent, but no detail about whether it was something sane...
If we spy, we could potentially validate it against the object we passed in through the http request

Reply:
If it's possible to avoid spies that would be good. A custom matcher (ie `argThat`) or a captor and then whatever assertThat's after would be much nicer.  Spies wind up coupling to the internals of what's being tested a lot.

Reply:
Yep, I've used an ArgumentCaptor 👍

Reply:
sorry yep that's what i mean by a spy - goodo :)